### DISCUSSION TITLE
Testing strategy

### DISCUSSION THREAD

**Discussion by ymatsiuk** [fluxcd/flux2#1039] [Q&A] [answered]
Would be awesome to know what kind of testing strategy is possible to prevent `master` from breaking.

Context:

- we use helm
- helm might break (forgot to update `values.yaml` or missing sneaky breaking changes)

Potential resolution:

- deploy `staging` namespace from every PR and report status back into github
- rollback `staging` to `master` once deployment succeeded
- once PR is accepted sync `staging` and `production` namespaces

Potential problems:

- `staging` must be pointing to `master` as the main source of truth, so the only feasible solution for this flow is to run `flux create ..` from actions

Any thoughts on this? How do you make sure that master always works?

PS: I know that flux is not going to apply changes is validation fails, but is there a way to validate PRs? Something like `flux validate ...` would be handy. Or even some kind of config field `validationBranch: stg*` which tells flux to pull the repo and run validation of resource from branches `stg`-prefixed against configured namespace. So we could configure webhook to report commit state back to PR.

**Comment by stefanprodan** [comment, marked as answer]
We test PRs by running flux in Kubernetes Kind, you can find an example here https://github.com/fluxcd/flux2-kustomize-helm-example#testing

> Any change to the Kubernetes manifests or to the repository structure should be validated in CI before a pull requests is merged into the main branch and synced on the cluster. This repository contains the following GitHub CI workflows:
> - the [test workflow](https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/.github/workflows/test.yaml) validates the Kubernetes manifests and Kustomize overlays with kubeval
> - the [e2e workflow](https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/.github/workflows/e2e.yaml) starts a Kubernetes cluster in CI and tests the staging setup by running Flux in Kubernetes Kind

**Reply by ymatsiuk** [reply]
Thanks for the fast reply. Checking

**Reply by tuxillo** [reply]
But would that approach work for production clusters like clusters in EKS?

Also there are things that when deployed, make API calls to retrieve data or make changes that you will probably not want in a test environment. For example, if you use external-dns, you don't want your DNS records to be created from the test cluster.

**Comment by tuxillo** [comment]
Any recommendation please?

**Reply by kingdonb** [reply]
If your production environment has stateful components that are important to keep isolated from the test environment, then you cannot make an end-to-end test. But you could make a test with mocks. You need to have a well-defined contract with the stateful components, so that you can abstract them out and do your end-to-end testing on the "everything else" parts of your deployment.

This is a challenging issue, and I've worked before in regulated scenarios where the only possible solution was to keep a replicated clone copy of a recent snapshot of prod, and to occasionally blow it away (refresh) in the dev environment, so that we could keep up to date with the developments in Prod without worry that someone is going to deploy a change in a stateful part that has some impact on our dev work.

I'm not sure what impact changing out the stateful components like load balancers or persistent volumes might have on your prod clusters, the advice for EKS is generally the same as the advice for every place else. Have multiple environments, and do the best you can to keep the non-prod environments in a reasonable state approximating the additional challenges that we're sure to encounter in prod. You still need disaster recovery plans, etc. They can be vetted in dev and staging environments, but they should eventually be tested in production even if there has been no outage, since that's the environment where the outage response will need to occur.

You should use dynamic provisioning, local path provisioner, etc. for those things that need a stateful component, and ideally there's still something meaningful you can determine from the e2e without making a complete replica of those private stateful parts (user db).

Sometimes, genuinely from an engineering perspective, the e2e testing really isn't complete without a more elaborate set of mocks and more complete modeling of those private stateful parts. Sometimes you really need to engage with production cloud APIs, albeit in a testing environment, in order for end to end testing scenarios to be useful and complete enough to be maximally beneficial as e2e's.

The Flux project itself has some quite elaborate e2e tests that have been further developed as part of the 2.0.0 release, to ensure the most complete coverage of our integration surfaces with the various cloud providers. Depending on your definition of e2e and your requirements, you may not want to spin up a temporary EKS cluster just to verify the configuration that you have internalized in the Git repo before sending it forward to production.

You can use a non-EKS cluster for e2e tests, I recommend vcluster or kind as two great options depending on whether you have a non-prod Kubernetes available to host your test workloads on, vcluster can make an excellent sandbox. But if your Git configuration internalizes EKS details and the EKS cluster itself is being managed from Git, then you maybe won't have been able to cover those details through any e2e that you build on such a vcluster, and the e2e suite may have a gap.

It's up to your own tolerance and risk management procedures whether this testing gap as a compromise is acceptable, I guess, or if it needs to be plugged with further testing. Likely your EKS clusters themselves are not changing very often, outside of scalers. HTH!

**Reply by tuxillo** [reply]
Thanks for the analysis!

Indeed, it's not a trivial matter since there are many factors to take in account. When I asked I was assuming cloning the Prod environment was out of question, because the cost and the time it would take to startup which, in my opinion, would hinder the development process.

We don't know what we're going to do it in the end but we'll probably produce a minimal sandbox to verify that the core apps reconciliation works and handle the rest of the issues in a environment to environment basis (dev to staging to prod).
